File: rust/src/solver.rs
use nalgebra::{DMatrix, DVector};

/// Solve steady-state occupation probabilities from rate equation 0 = WP.
/// Uses augmented system: replace last row with normalization sum(P)=1,
/// solve via QR decomposition, clamp negatives, renormalize.
pub fn solve_steady_state(w: &[f64], n: usize) -> Vec<f64> {
    let dim = 2 * n;
    if dim == 0 {
        return vec![];
    }

    // Build augmented matrix: copy W, replace last row with 1s
    let a = DMatrix::from_fn(dim, dim, |i, j| {
        if i == dim - 1 {
            1.0
        } else {
            w[i * dim + j]
        }
    });

    let mut d = DVector::zeros(dim);
    d[dim - 1] = 1.0;

    // Solve via QR decomposition
    let qr = a.qr();
    let x = match qr.solve(&d) {
        Some(sol) => sol,
        None => return vec![0.0; dim],
    };

    // Clamp negatives, renormalize
    let mut p: Vec<f64> = x
        .iter()
        .map(|&v| if v.is_finite() && v >= 0.0 { v } else { 0.0 })
        .collect();

    let sum: f64 = p.iter().sum();
    if sum > 0.0 {
        for v in &mut p {
            *v /= sum;
        }
    } else {
        p = vec![0.0; dim];
        p[dim - 1] = 1.0;
    }

    p
}

File: cpp/src/solver.cpp
#include "solver.hpp"

#include <cmath>
#include <Eigen/Dense>

namespace fc {

void solve_steady_state(double* W, double* P, int N) {
    int dim = 2 * N;
    if (dim <= 0) {
        return;
    }

    Eigen::MatrixXd A(dim, dim);
    for (int i = 0; i < dim; ++i) {
        for (int j = 0; j < dim; ++j) {
            A(i, j) = W[i * dim + j];
        }
    }

    A.row(dim - 1).setOnes();

    Eigen::VectorXd rhs = Eigen::VectorXd::Zero(dim);
    rhs(dim - 1) = 1.0;

    Eigen::VectorXd x = A.householderQr().solve(rhs);

    double sum = 0.0;
    for (int i = 0; i < dim; ++i) {
        double v = x(i);
        if (!std::isfinite(v) || v < 0.0) {
            v = 0.0;
        }
        P[i] = v;
        sum += v;
    }

    if (sum > 0.0) {
        for (int i = 0; i < dim; ++i) {
            P[i] /= sum;
        }
    } else {
        for (int i = 0; i < dim; ++i) {
            P[i] = 0.0;
        }
        P[dim - 1] = 1.0;
    }
}

}
